docs/public-async-nio-apis.md

```diff
@@ -0,0 +1,1162 @@
+    ///   - shouldUpgrade: A callback that determines whether the websocket request should be
+    ///         upgraded. This callback is responsible for creating a `HTTPHeaders` object with
+    ///         any headers that it needs on the response *except for* the `Upgrade`, `Connection`,
+    ///         and `Sec-WebSocket-Accept` headers, which this upgrader will handle. Should return
+    ///         an `EventLoopFuture` containing `nil` if the upgrade should be refused.
+    ///   - enableAutomaticErrorHandling: A function that will be called once the upgrade response is
+    ///         flushed, and that is expected to mutate the `Channel` appropriately to handle the
+    ///         websocket protocol. This only needs to add the user handlers: the
+    ///         `WebSocketFrameEncoder` and `WebSocketFrameDecoder` will have been added to the
+    ///         pipeline automatically.
+    public init(maxFrameSize: Int = 1 << 14, enableAutomaticErrorHandling: Bool = true, shouldUpgrade: @escaping @Sendable (NIOCore.Channel, NIOHTTP1.HTTPRequestHead) -> NIOCore.EventLoopFuture<NIOHTTP1.HTTPHeaders?>, upgradePipelineHandler: @escaping @Sendable (NIOCore.Channel, NIOHTTP1.HTTPRequestHead) -> NIOCore.EventLoopFuture<UpgradeResult>)
+
+    /// Builds the upgrade response headers. Should return any headers that need to be supplied to the client
+    /// in the 101 Switching Protocols response. If upgrade cannot proceed for any reason, this function should
+    /// return a failed future.
+    public func buildUpgradeResponse(channel: NIOCore.Channel, upgradeRequest: NIOHTTP1.HTTPRequestHead, initialResponseHeaders: NIOHTTP1.HTTPHeaders) -> NIOCore.EventLoopFuture<NIOHTTP1.HTTPHeaders>
+
+    /// Called when the upgrade response has been flushed. At this time it is safe to mutate the channel pipeline
+    /// to add whatever channel handlers are required. Until the returned `EventLoopFuture` succeeds, all received
+    /// data will be buffered.
+    public func upgrade(channel: NIOCore.Channel, upgradeRequest: NIOHTTP1.HTTPRequestHead) -> NIOCore.EventLoopFuture<UpgradeResult>
+}
+```
+
+### ALPN
+
+```swift
+/// The result of protocol negotiation.
+public struct NIOProtocolNegotiationResult<NegotiationResult> where NegotiationResult : Sendable {
+
+    /// Intializes a new ``NIOProtocolNegotiationResult`` with a final result.
+    ///
+    /// - Parameter result: The final result of protocol negotiation.
+    public init(result: NegotiationResult)
+
+    /// Intializes a new ``NIOProtocolNegotiationResult`` with a deferred result.
+    ///
+    /// - Parameter deferredResult: The deferred result.
+    public init(deferredResult: NIOCore.EventLoopFuture<NIOCore.NIOProtocolNegotiationResult<NegotiationResult>>)
+
+    @available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)
+    @available(*, deprecated, renamed: "getResult")
+    public func waitForFinalResult() async throws -> NegotiationResult
```

Comment:
If we're moving this to public API should we delete the deprecated code?

Reply:
Ah good catch this should be removed. Thanks!